Client/game_sa/CColModelSA.h

```diff
@@ -58,19 +63,123 @@ typedef struct
     WORD                 numColBoxes;
     WORD                 numColTriangles;
     BYTE                 ucNumWheels;
-    BYTE                 pad3;
+    BYTE                 m_nFlags;
     CColSphereSA*        pColSpheres;
     CColBoxSA*           pColBoxes;
     void*                pSuspensionLines;
-    void*                pUnknown;
+    CompressedVector*    pVertices;
     CColTriangleSA*      pColTriangles;
     CColTrianglePlaneSA* pColTrianglePlanes;
+    unsigned int         m_nNumShadowTriangles;
+    unsigned int         m_nNumShadowVertices;
+    CompressedVector*    m_pShadowVertices;
+    CColTriangleSA*      m_pShadowTriangles;
+
+    std::map<ushort, CompressedVector> getAllVertices() const
+    {
+        std::map<ushort, CompressedVector> vertices;
+        for (uint i = 0; numColTriangles > i; i++)
+        {
+            vertices[pColTriangles[i].vertex[0]] = pVertices[pColTriangles[i].vertex[0]];
+            vertices[pColTriangles[i].vertex[1]] = pVertices[pColTriangles[i].vertex[1]];
+            vertices[pColTriangles[i].vertex[2]] = pVertices[pColTriangles[i].vertex[2]];
+        }
+        return vertices;
+    }
+
+    size_t getNumVertices() const
+    {
+        std::map<ushort, bool> vertices;
+        for (uint i = 0; numColTriangles > i; i++)
+        {
+            vertices[pColTriangles[i].vertex[0]] = true;
+            vertices[pColTriangles[i].vertex[1]] = true;
+            vertices[pColTriangles[i].vertex[2]] = true;
+        }
+        return vertices.size();
+    }
+
+    bool isValidIndex(char eShape, ushort usIndex, ushort numVertices = 0) const
+    {
+        switch (eShape)
+        {
+            case 0:
+                return (usIndex >= 0 && usIndex < numColBoxes);
+                break;
+            case 1:
+                return (usIndex >= 0 && usIndex < numColSpheres);
+                break;
+            case 2:
+                return (usIndex >= 0 && usIndex < numColTriangles);
+                break;
+            case 3:
+                return (usIndex >= 0 && usIndex < numVertices);
+                break;
+        }
+        return false;
+    }
+
+    std::vector<ushort> getTrianglesByVertex(ushort usVertex) const
+    {
+        std::vector<ushort> vecTriangles;
+        CColTriangleSA      colTriangle;
+        for (ushort i = 0; i < numColTriangles; i++)
+        {
+            colTriangle = pColTriangles[i];
+            if (colTriangle.vertex[0] == usVertex || colTriangle.vertex[1] == usVertex || colTriangle.vertex[2] == usVertex)
+                vecTriangles.push_back(i);
+        }
+        return vecTriangles;
+    }
+
+    std::map<ushort, ushort> getVerticesUsage() const
+    {
+        std::map<ushort, ushort> verticesUsage;
+        for (ushort i = 0; i < numColTriangles; i++)
+            for (char cVertex = 0; cVertex < 3; cVertex++)
+                verticesUsage[pColTriangles[i].vertex[cVertex]]++;
+        return verticesUsage;
+    }
```

Comment:
Same issue here again.

Reply:
@saml1er isnt that RVOd? I woudnt be so concerned about it.

Reply:
Or he even could `inline` it.

Reply:
inline isn't guaranteed, but you're right about NRVO. It's guaranteed in C++. This code is okay.